### PR TITLE
docs(scanner): Create ADR to adopt ADRs in Scanner

### PR DESCRIPTION
**Pull request by jvdm**
## Checklist

- [x] Investigated and inspected CI test results

(N/A):

- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

None.

### CODE REVIEW

**Review by RTann — APPROVED** [review]

File: scanner/decisions/0000-adopt-adrs.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+4. **ADRs are public and local by default:** ADRs will be kept in a specific directory within the code repo to correlate decisions with code changes and make them widely available, even to external parties. The ADR will be stored in a private location when the context is sensitive (e.g., security, customer information).
+5. **ADR template:** ADRs will use a well-defined template.
+
+## Consequences
+
+- **Additional overhead:** ADRs are now required when a few slack threads would be OK. That creates some overhead to long-term discussions and decisions that did not exist before.
+- **Improved Clarity and Velocity:** This approach will bring clear communication, faster onboarding time, a better context for discussions, and reduce unnecessary back-and-forths in other contexts (e.g., PR).
+- **Transparency and Accessibility:** By making ADRs accessible, internal team members and external stakeholders will benefit from increased visibility into the decision-making process.
```
Comment (dcaravel) [inline scanner/decisions/0000-adopt-adrs.md:32]: FWIW, in my past we went down a similar route and learned that content in GitHub was not as easily consumed by personas with a non-technical background. This resulted in less engagement in some circumstances.

**Review by dcaravel — APPROVED** [review]

File: scanner/decisions/0000-adopt-adrs.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+- **Unclear Boundaries:** There's no clear framework to help us understand when the overhead of design documents is worth it.
+
+## Decision
+
+1. **ADRs for long-term impact decisions:** ADRs will be used to document decisions that have a long-term impact on the project. This process is considered more mature and reliable than other methods (e.g., slack threads).  We will continue to use informal methods for trivial decisions.
+2. **ADRs for smaller and larger decisions:** ADRs can be used for smaller and larger decisions, whether public or not, depending on the context.
+3. **Design docs to aid complex contexts:** Design documents can be used when the context is ambiguous and requires an in-depth analysis of options available. ADRs will still be used to formally record the decision.[^1]
+4. **ADRs are public and local by default:** ADRs will be kept in a specific directory within the code repo to correlate decisions with code changes and make them widely available, even to external parties. The ADR will be stored in a private location when the context is sensitive (e.g., security, customer information).
+5. **ADR template:** ADRs will use a well-defined template.
```
Comment (dcaravel) [inline scanner/decisions/0000-adopt-adrs.md:26]: Will this template include where to store binary assets?  For example, if the ADR includes a diagram / picture I would expect anyone with edit access to the ADR to also be able to edit the diagrams/pictures/etc.  Or perhaps we settle on using only assets that can be represented 'as code', ie: diagrams via [mermaidjs](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/)

**Review by dcaravel — APPROVED** [review]

File: scanner/decisions/0000-adopt-adrs.md (hunk 2 above)
Comment (jvdm) [reply scanner/decisions/0000-adopt-adrs.md:26]: I haven't considered that yet. I agree it's important to make sure diagrams are editable, but I'm not sure which platform would be best. Instead of enforcing, I believe preferring code-based diagrams would be better. Additionally, there is a restriction on the size of files in the repo (50K, I think), which will require us to use simple and focused diagrams. We can certainly improve things as we go and document the options in the `template.md` file.

**Review by daynewlee — APPROVED** [review]

File: scanner/decisions/0000-adopt-adrs.md (hunk 1 above)
Comment (jvdm) [reply scanner/decisions/0000-adopt-adrs.md:32]: I wonder if that was a specific problem related to your previous context. I mean, could you please clarify the role of non-technical stakeholders in your case? I am asking because it might be an oversimplification to assume non-technical folks can't navigate GitHub, as many of them are capable of doing so. If only a minority of our non-technical audience (customers, support, community, etc.) cannot navigate GitHub, it might not justify moving ADRs away. Do you have any suggestions for other platforms or formats that might address the issue you are raising without the drawbacks of moving away from the repo?

Comment (dcaravel) [reply scanner/decisions/0000-adopt-adrs.md:32]: Not saying move away, just something we'll need to pay attention to.

A few examples:
- **Access**
  - Easy to grant access to a google doc, not so much for a private GH repo, or confluence - the latter(s) requiring tickets or tracking down admins.
- **Discussion/Comments**
  - During a PR simple to leave comments / have a discussion
  - When no PR is active, not as easy to have a discussion, usually happens in another medium like slack, gmeet, etc. and then a PR created
  - GitHub does have a discussions feature that perhaps could be enabled to solve this.
- **Contributing/Suggesting**
  - Standard github workflow: clone/fork, edit, create PR, discuss, merge can be complex for some.
  - Compared to gdocs or confluence: make a change/suggestion/comment directly.
- **Tasks/Followups**
  - In gdocs and confluence can create tasks and assign them to folks, notifications followups and tracking are handled
  - In GH can tag someone, but the tracking and followup is manual.

Regarding personas, anyone without hands-on GH experience - thinking of product managers, non-engineering managers, or perhaps officers that aren't hands on.

File: scanner/decisions/0000-adopt-adrs.md (hunk 2 above)
Comment (dcaravel) [reply scanner/decisions/0000-adopt-adrs.md:26]: Thinking about this more, perhaps a link to lucid, miro, whatever if thats where the diagram was created would be sufficient.  Ideally placing them into a shared folder/group mirroring GH repo access.

That max size must be adjustable - In my past we've stored videos in GH that were 500 MiB+

File: scanner/decisions/0000-adopt-adrs.md (hunk 1 above)
Comment (jvdm) [reply scanner/decisions/0000-adopt-adrs.md:32]: > Not saying move away, just something we'll need to pay attention to.

I know, I am assuming that given you approved this.

I will reply with my personal view of things.

> Access

That doesn't apply to public github repositories, right? For private, assuming we are using it to control access to sensitive information, I think the extra step and control over who has access is desirable.

>  Discussion/Comments: When no PR is active, not as easy to have a discussion, usually happens in another medium like slack, gmeet, etc. and then a PR created / GitHub does have a discussions feature that perhaps could be enabled to solve this.

In the context of ADRs, we commit to an architectural decision while the PR is open by discussing using the PR platform. After merging, a new decision is required to change it. If you want to discuss a decision and no PR is open, open a new ADR PR. Short-lived decisions, discoveries, design docs, and other discussion types, can happen elsewhere where this type of workflow is not relevant.

> Contributing/Suggesting: Standard github workflow: clone/fork, edit, create PR, discuss, merge can be complex for some.

GitHub's web platform enables users to add and edit markdown files without the need to clone locally using git. Markdown editing in pull requests is similar to other collaboration features in the platform (e.g., comments). It probably lacks rich editing, but that is OK and to some extent maybe even desirable for ADRs.

> Contributing/Suggesting: Compared to gdocs or confluence: make a change/suggestion/comment directly.

Making comments and suggestions in GitHub's PR review UI is more robust than in gdocs. Gdocs offers a simplified interface, without an easily accessible history, and poor thread support. It also makes it difficult to follow and track the conversation across the changes in the document's history.

In the context of ADRs, allowing changes in line might be an anti-pattern. It is easier, but being a record, the final text is essential and should not change without everyone knowing why and how it changed. Especially during the review and discussion. In gdocs such changes might happen in the doc and you might not capture it, not be aware of it.

> Tasks/Followups: In GH can tag someone, but the tracking and followup is manual.

In GitHub, it's possible to link issues to PRs and assign them to people, generating notifications. However, we primarily use JIRA instead of tracking GitHub issues. Integrations might be available for that, I suppose.

TL;DR: Public GitHub repos tackle the access issue, and for private ones, tight access control is a plus. For ADRs, PRs serve as the discussion platform before a decision is finalized, and anyone can initiate a new PR for further talks. GitHub's web UI makes markdown edits straightforward, and its PR review system is robust, ensuring changes are trackable and discussions are captured. It's more structured than gdocs, which is key for decision records.

Comment (jvdm) [reply scanner/decisions/0000-adopt-adrs.md:32]: > Regarding personas, anyone without hands-on GH experience - thinking of product managers, non-engineering managers, or perhaps officers that aren't hands on.

Thanks for clarifying the personas. But my question was focused on how much these roles or personas in your previous community compare to our context now regarding familiarity with GitHub PRs, to the point that would prevent today's from accessing the docs.

File: scanner/decisions/0000-adopt-adrs.md (hunk 2 above)
Comment (jvdm) [reply scanner/decisions/0000-adopt-adrs.md:26]: As long as anyone can access the link and see the diagram, I would be OK.

> That max size must be adjustable

FYI, I think that's part of our post-commit checks, but I am not sure.
